# rMETL: 
rMETL - realignment-based Mobile Element insertion detection Tool for Long read

---
### Getting Start
		       _  ___  _   _____   _______   _
		 _ _  | ^_   _^ | |  ___| |__   __| | |
		| ^_| | | | | | | | |__      | |    | |
		| |   | | | | | | |  __|     | |    | |
		| |   | | | | | | | |___     | |    | |___
		|_|   |_| |_| |_| |_____|    |_|    |_____|
     
	
	$ git clone https://github.com/hitbc/rMETL.git (git clone https://github.com/tjiangHIT/rMETL.git)
	$ cd rMETL/
	$ bash INSTALL.sh
	$ ./rMETL.py

---	
### Introduction



---
### Simulated datasets

The simulated datasets use for benchmarking are available at: https://drive.google.com/open?id=1ujV2C8e1PNAVhSkh9vKtjWLdG_OHcH-k

---
### Memory usage

The memory usage of rMETL can fit the configurations of most modern servers and workstations.
Its peak memory footprint is about 16.7G Gigabytes (default setting), on a server with Intel Xeon CPU at 2.00 GHz, 1 Terabytes RAM running Linux Ubuntu 14.04. These reads were aligned to human reference genome hs37d5.

---
### Dependences
	
	1. pysam
	2. Biopython
	3. ngmlr
	4. samtools

	Python version 2.7

---
### Installation

Current version of rMETL needs to be run on Linux operating system.
The source code is written in python, and can be directly download from: https://github.com/hitbc/rMETL 
A mirror is also in: https://github.com/tjiangHIT/rMETL
The INSTALL.sh is attached. Use the make command for generating the executable file.

---
### Synopsis



---
### Commands and options




---
### Reference


---
### Contact
For advising, bug reporting and requiring help, please contact ydwang@hit.edu.cn or tjiang@hit.edu.cn
