# nTED: 
non-reference Transposable Elements detection with long sequencing reads

---
### Getting Start
		         _______   _____   ___
		 _ ___  |__   __| |  ___| |  _ \
		| ^_  \    | |    | |___  | | \ |
		| | | |    | |    |  ___| | | | |
		| | | |    | |    | |___  | |_/ |
		|_| |_|    |_|    |_____| |____/
     
	
	$ git clone https://github.com/hitbc/nTED.git (git clone https://github.com/tjiangHIT/nTED.git)
	$ cd nTED/
	$ bash INSTALL.sh
	$ ./nTED.py

---	
### Introduction



---
### Simulated datasets

The simulated datasets use for benchmarking are available at: https://drive.google.com/open?id=0Bxxw-cTRcGuHUVVUS01BeW5aRXc

---
### Memory usage

The memory usage of nTED can fit the configurations of most modern servers and workstations.
Its peak memory footprint is about 16.7G Gigabytes (default setting), on a server with Intel Xeon CPU at 2.00 GHz, 1 Terabytes RAM running Linux Ubuntu 14.04. These reads were aligned to human reference genome hs37d5.

---
### Dependences
	
	1. pysam
	2. Biopython
	3. ngmlr
	4. samtools

	Python version 2.7

---
### Installation

Current version of nTED needs to be run on Linux operating system.
The source code is written in python, and can be directly download from: https://github.com/hitbc/nTED 
A mirror is also in: https://github.com/tjiangHIT/nTED
The INSTALL.sh is attached. Use the make command for generating the executable file.

---
### Synopsis



---
### Commands and options




---
### Reference


---
### Contact
For advising, bug reporting and requiring help, please contact ydwang@hit.edu.cn or tjiang@hit.edu.cn
